Add "--network-opt" option as a way to pass arbitrary options to network drivers

axqx opened the pull request:
- Fixes https://github.com/docker/libnetwork/issues/910
  The docker plugin model has made it easier to extend docker with custom
  drivers that can be maintained independently by third parties. However,
  as reported by issue https://github.com/docker/libnetwork/issues/910
  a way of passing arbitrary parameters from docker client to remote
  driver is needed.
  Many use cases require passing network configuration to a network driver
  that might vary from container to container and might only be known at
  container creation time. Examples are configuring the MTU or queue
  length of a network interface or passing ip aliases, routes or iptables
  rules for containers. This patch proposes adding a new --net-opt option
  to docker to allow passing arbitrary information to network drivers.
  net-opt options are used with docker run command and are directly copied
  to the container's NetworkSettings and then passed to the network driver
  on endpoint create.

Signed-off-by: Alina Quereilhac alina@medallia.com

Commit:
Add --network-opt and --ipam-opt as a way to pass arbitrary options to network drivers
- Fixes #910

The docker plugin model has made it easier to extend docker with custom
drivers that can be maintained independently by third parties. However,
as reported by issue https://github.com/docker/libnetwork/issues/910
a way of passing arbitrary parameters from docker client to remote
driver is needed.
Many use cases require passing network configuration to a network driver
that might vary from container to container and might only be known at
container creation time. Examples are configuring the MTU or queue
length of a network interface or passing ip aliases, routes or iptables
rules for containers. This patch proposes adding a new --net-opt option
to docker to allow passing arbitrary information to network drivers.
net-opt options are used with docker run command and are directly copied
to the container's NetworkSettings and then passed to the network driver
on endpoint create.

Signed-off-by: Alina Quereilhac <alina@medallia.com>

mavenugo commented:
@axqx thanks for the PR. am generally 👍 for providing such flexibility for users to make use of the driver capabilities. I initially tried to get this via labels (https://github.com/docker/docker/pull/13476), but was not accepted for good reason.

If we have an explicit option (I will change it to `--network-opt` to be consistent with other flags), that will make it clear. I am in general positive with this design.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "net-opt" git@github.com:medallia/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

axqx commented:
@mavenugo, thank you for your feedback. I added the change you suggested (--net-opt to --network-opt) in commit 4a5768604535a867d16c8509970eece1e6b42a6d.

thaJeztah commented:
@axqx don't forget to update the API changes; [docs/reference/api/docker_remote_api.md#v125-api-changes](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes), and the networks section in the API reference itself [docs/reference/api/docker_remote_api_v1.25.md#35-networks](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.25.md#35-networks)

axqx commented:
Thank you @thaJeztah. The API documentation is now updated in commit  d24405599bb314da4de3c3a509bc7d607e463c24

mavenugo commented:
@axqx while we are at this, shall we expand this a little more to make it more effective ?

* add `--ipam-opt` to be passed to IPAM drivers/plugins
* add `--network-opt` and `--ipam-opt` to `docker service` command as well. The end-to-end integration will require changes in swarmkit project as well cc @aluzzardi . I had a brief chat with @aluzzardi regarding adding such an option in swarmkit to benefit from the work done here.

@icecrime @mrjana given we have multiple requests for such a support and we couldn't get https://github.com/docker/docker/pull/13476 in ... can you share your views on adding such an option ?

jc-m commented:
@mavenugo Don't you think that we should make this PR as atomic as possible to avoid blocking forward progress ?

axqx commented:
@mavenugo , I agree with @jc-m , while adding an --ipam-opt option might also be useful, I think this should probably be addressed in a different PR. I would rather keep this PR to address only one issue and one new functionality.

icecrime commented:
@mavenugo Seems fine by me, I'm happy as long as we're not using labels as the vehicle for arbitrary driver-specific options. A couple things that we maybe need to look into:

- Is there going to be any relationship between the `--network-opt` flag to `run` and the `--opt` flag to `network create`? I can easily see some confusion if different drivers have different conventions (e.g., anything can be specified at run time versus different subsets of options for `run` and `network create`).
- What about the consistency with volumes?

mavenugo commented:
@axqx @jc-m yes. agreed. We can move the `docker service` requirement out of this PR and atleast try to cover both network and ipam options. It will be so annoying otherwise for users who have been asking for such feature using labels. Now that we have a proposal with a driver specific option, lets try to knock these 2 down in 1 go. WDYT ?

mavenugo commented:
@icecrime thanks for the confirmation.

>> Is there going to be any relationship between the --network-opt flag to run and the --opt flag to network create? I can easily see some confusion if different drivers have different conventions (e.g., anything can be specified at run time versus different subsets of options for run and network create).

These options are directly related to CNM objects of Network and Endpoint. Driver option for network goes via `--opt` in `network create`, while the driver options for endpoint goes via this proposed flag. And since we allow just 1 network in `docker run` and `network connect`, we can associate which `--network-opt` is for which network-endpoint combination and I guess that is the intended behavior.

>> What about the consistency with volumes?

Good question. Is there such a requirement ? @cpuguy83

mavenugo commented:
@axqx also, this should address the `docker network connect` use-case as well. We cannot have it just in `docker run` and not in `docker network connect`.

cpuguy83 commented:
Volumes have parity via `--mount`.
We could introduce a `volume-opt`, but I think it would be overly complex when someone could/should just use `--mount`.

mauri commented:
+1 on adding these options to `docker network connect`. I'm not sure about introducing IPAM here, today they are separated concerns for a reason. I don't think that we should just piggyback unrelated changes.

mavenugo commented:
@mauri it is not unrelated. They are closely related concepts with different hook points. A good example of this is the `--ip` option that we introduced in `docker run` and `--subnet` option in `docker network create`. I do believe the `--ipam-opt` belong here.

icecrime commented:
@mavenugo Playing devil's advocate here:

> These options are directly related to CNM objects of Network and Endpoint. Driver option for network goes via --opt in network create, while the driver options for endpoint goes via this proposed flag. And since we allow just 1 network in docker run and network connect, we can associate which --network-opt is for which network-endpoint combination and I guess that is the intended behavior.

Is it `docker run --endpoint-opt` then? :innocent:

mavenugo commented:
@icecrime 

>> Is it docker run --endpoint-opt then? 😇

😄  True. But the user really don't need to worry about that. Since this is a per-container option for that "network", `--network-opt` does indicate endpoint option. (Since we allow only 1 network per container operation).

icecrime commented:
Fair enough :wink: I think it makes sense when properly explained: this configures driver-specific options for attaching that particular container to the network.

Design LGTM :+1:

axqx commented:
@mavnugo, thanks for pointing out the docker network connect use-case, indeed it needs to be addressed here. Regarding the --ipam-opt, it is certainly a related feature but implementation wise is independent from this change, so I really think it belongs to a different PR. I believe that separating independent functionality into different PRs allows to progress faster towards successfully integrating new features.

mavenugo commented:
@axqx if you purely look at it from a PR pov, then yes, I agree to have simple dedicated PRs. But I look also at the UX and want to make sure it is covering all the bases. We don't want the users to assume `network-opt` is used for both network and ipam drivers. Also, logically it doesn't make sense to have a driver option for network and not for ipam drivers.

My concern also come from release engineering, If you are targeting this for the upcoming 1.13 release, then we don't want to go with just this option because it will impact the UX as mentioned above. Unless we can cram in another PR for --ipam-opt prior to 1.13.
If you are targeting it for 1.14, then we can continue to review it post 1.13 code-freeze so that we will have enough time for the release to carry 2 PRs for network-opt and ipam-opt to go through the review and merge cycle.

Hope you understand my concerns in just having the network-opt without ipam-opt. am all ears if other maintainers think otherwise.

mavenugo commented:
@axqx any updates ? We have to make sure if this needs to be in 1.13 milestone or not.

axqx commented:
@mavenugo , sorry, not much progress yet. Unfortunately I will not have bandwidth to continue work on this until next week. Seems target date for 1.13 code freeze was 10/27/2016, so I don't think I will manage to have this feature implemented for 1.13 milestone :(.

mavenugo commented:
@axqx 

> Seems target date for 1.13 code freeze was 10/27/2016

No. https://github.com/docker/docker/wiki is out-of-date and we have to fix that. 
The Code-freeze has moved by a couple of weeks to 11/10/2016 and you still have the time to get this in prior to that if you prefer. Sooner the better to give enough time for review and merge.

axqx commented:
@mavenugo , good to know. I'll dedicate time to this as early as possible next week and try to send updates early to see if it can make it to 1.13 milestone.

axqx commented:
@mavenugo, added the requested changes: --network-opt is now added to docker network connect and --ipam-opt to docker network run and docker network connect to be passed to IPAM driver on endpoint create.

mcastelino commented:
@axqx @mavenugo As docker now supports alternate OCI compliant runtimes. Does it make sense to build on this PR to provide meta data from the runtime to the network plugin, so that the network plugins can setup the networking interfaces more optimally for runtimes other than runc. 

The information can be provided to the CNM by the runtime at the invocation of the network pre-hook. 

In our case we have a VM based OCI compliant runtime https://github.com/01org/cc-oci-runtime. Here if the network plugin could create a tap interface vs a veth interface based on the meta data sent back from the runtime, it would make the networking flow significantly more efficient.

jc-m commented:
@mcastelino You can build on top of this PR, as soon as it's integrated. But let's not expand the scope any more :-)

lxpollitt commented:
@mavenugo, @axqx It would be great to get this merged in sometime soon.  Do you think that is likely to happen?

axqx commented:
@lxpollitt, @mavenugo, please let me know if there are any improvements I should do to move this PR forward. Also, I would like to update the API changes to the new remote API version (v1.26) but I don't find docs/reference/api/docker_remote_api_v1.25.md or docs/reference/api/docker_remote_api_v1.26.md in master branch (they are both referenced by docs/api/version-history.md).

mauri commented:
Hi @thaJeztah, the API docs seems to have moved; where can we find them to move this forward? Thanks!

thaJeztah commented:
API docs are now generated from the swagger.yml; see https://github.com/docker/docker/tree/master/api, and changelog for the API has moved to https://github.com/docker/docker/blob/master/docs/api/version-history.md

mavenugo commented:
Ping @aboch

mavenugo commented:
@axqx can u pls rebase ?

aboch commented on api/types/network/network.go:
```diff
@@ -23,9 +23,10 @@ type IPAMConfig struct {
+	IPv4Address  string            `json:",omitempty"`
+	IPv6Address  string            `json:",omitempty"`
+	LinkLocalIPs []string          `json:",omitempty"`
+	IpamOpts     map[string]string `json:"IPAMOpts,omitempty"`
```
Given it is a member of `EndpointIPAMConfig`, I'd simply call this `Options`. Also in the json tag.

aboch commented on cli/command/network/connect.go:
```diff
@@ -43,6 +47,8 @@ func newConnectCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.StringSliceVar(&opts.aliases, "alias", []string{}, "Add network-scoped alias for the container")
 	flags.StringSliceVar(&opts.linklocalips, "link-local-ip", []string{}, "Add a link-local address for the container")
+	flags.Var(&opts.networkOpts, "network-opt", "Set network driver option for a container endpoint")
+	flags.Var(&opts.ipamOpts, "ipam-opt", "Set IPAM driver option for a container endpoint")
```
To conform to the help line for the network driver options in the network create command (in network_create.go), please use the plural `options`

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "net-opt" git@github.com:medallia/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353461456
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

axqx commented:
@mavenugo, @aboch, thank you for the comments. Branch is rebased and comments addressed in last commit.

thaJeztah commented:
@axqx looks like it needs a `gofmt`;

```
01:22:48 These files are not properly gofmt'd:
01:22:48  - runconfig/opts/parse.go
01:22:48 
01:22:48 Please reformat the above files using "gofmt -s -w" and commit the result.
```

axqx commented:
@thaJeztah, thank you. The file is now reformatted.

LK4D4 commented:
I see that design was reviewed for this PR.
@axqx would you mind to rebase pls?
ping @mavenugo @aboch @sanimej for code-review also

axqx commented:
@LK4D4, sure, rebased.

mavenugo commented:
Thanks @axqx 

Could you write a test case for this and make sure a plugin infact receives these options e2e ?
We have a few plugin related test cases in  integration-cli/docker_cli_network_unix_test.go such as `TestDockerNetworkDriverOptions`.

thaJeztah commented:
Thinking about this; should `--network-opt` be separate from `--network`? We've made that mistake with `--volume-driver`, and because of that, only a single driver can be specified on `docker run`. The advanced (`--mount`) syntax that's present on `docker service create` resolves that, but is not available yet for `docker run`.

Would it make sense to have an advanced syntax for `--network` that allows specifying these options _per-network_? This would also allow specifying multiple networks on `docker run` (which we decided not to do, because per-network options may be needed), e.g.;


    --network name=my-net,opt=foo=bar,opt=foo2=baz


I haven't checked in-depth what kind of options are passed here, but thought to mention this, before it's too late :smile:

mavenugo commented:
@thaJeztah yes you are correct. The only reason why we have the issue of not able to support multiple `--network` in `docker run` is this particular issue of not able to wrap up all the network options in a group as you mentioned.

thaJeztah commented:
@mavenugo yes, so for that reason, I just thought "we have the new advanced syntax now". Using that for these options would solve that problem, and allow users to specify multiple networks on `docker run` as well.

m-kostrzewa commented:
@axqx awesome work! I'm writing a network driver right now and this feature would be very useful.

While I agree that having syntax that handles multiple networks would be great, I tend to agree with @jc-m more: scope of this PR was already expaned, I don't think it should be expanded any further. 

I'm sure @lxpollitt would love to get hands on this feature ASAP as well.

vdemeester commented:
ping @axqx 👼 🙏

axqx commented:
@m-kostrzewa, thanks!

@vdemeester, @mavenugo, @thaJeztah, sorry for the delay. Will try to work on the comments asap.

mavenugo commented:
@m-kostrzewa @axqx apologies on expanding the scope of this PR. We are between rock and hard place here. We recognize that the option is useful for network drivers. But we have made mistakes with the UX/API in the past which makes it hard to provide any reasonable compatibility. So as @thaJeztah suggested, I think it is better to start investing on the advanced `--network` syntax and use this as the first entry to it (along side network name). Let us not expand the scope of this PR beyond that and we will take up additional PRs to move other network specific options such as `-network-alias` and others into the advanced format which will let us support multiple networks in `docker run` and `docker service` together.

axqx commented:
@vdemeester, @mavenugo, @thaJeztah, added some integration tests and will work on the advanced --network syntax next.

thaJeztah commented:
Thanks @axqx - let me put this PR back in "design review" pending those changes 👍

Commit:
Add advanced --network and --ipam syntax for docker run, with network and ipam driver options

Syntax example:

docker run --network "name=network-name,key1=val1,key2=val2" \
	--ipam "name=network-name,key1=val1,key2=val2" --ip 10.1.0.2 alpine sh

Signed-off-by: Alina Quereilhac <alina@medallia.com>

axqx commented:
@mavenugo, @aboch, @sanimej, @thaJeztah, the PR is now updated with the advanced --network/ipam syntax for docker run. Could you please take a look at the changes?

Some usage examples:

```
docker run -d --name test busybox top # same as now, will use default network

docker run -d --name test --net my-net busybox top # same as now, will use my-net network

docker run -d --name test --net "name=my-net" busybox top # will use my-net network

docker run -d --name test --net "name=my-net,key1=val1,key2=val2" busybox top # advanced syntax, will use my-net passing options

docker run -d --name test --net "name=my-net,key1=val1,key2=val2" --ipam "name=my-net,key3=val3,key4=val4" busybox top # advanced syntax, will use my-net passing options to network and ipam drivers
```

geisbruch commented:
Hi, are you planning include this feature in the next release? It's very useful for custom driver. Today we should create many different networks to be hable to configure different ipam options on a custom driver.

mavenugo commented:
@axqx thanks for your effort on this. 

>> --net "name=my-net,key1=val1,key2=val2"

This doesn't capture the requirement well. Since there are many more network specific options other than name and driver opts. PTAL https://github.com/docker/docker/issues/31964 for an idea of what the proposal will look like.

Ofcourse for this PR, you are worried only about `--opt` and `--ipam-opt`. But as a generic UX, we have to cover all the network specific options under this new CSV format.
cc @thaJeztah . How realistic is it for us to get this all wrapped up for 17.05 ?

thaJeztah commented:
@mavenugo good question, IIUC, changes will all be in the CLI/client, so could be realistic to get it in if a maintainer or contributor is able to reserve some time to implement it.

thaJeztah commented:
Let me close this one for now, pending the changes for the `--network` flag (https://github.com/docker/docker/issues/31964). But @axqx if you have time and want to work on that, let me know! 

Happy to reopen if you want to start working on this again

axqx commented:
@thaJeztah, thanks for the update. Unfortunately I don't have time now to work on https://github.com/docker/docker/issues/31964, so I will wait until a stable form of the --network flag is achieved before resuming work on passing network options to drivers.
